### ISSUE TITLE
Using org-capture native expansion inside snippet

### ISSUE THREAD

**Issue by yantar92** [ag91/ya-org-capture#1]
I have tried the following setup:

```
(asoc-put! org-capture-templates
	   "E"
           (cdar (doct
		  '("Note"
		    :keys "E"
                    :type entry
                    :file "~/Org/inbox.org"
                    :clock-in t
                    :clock-resume t
                    :template
                    ("%(ya-org-capture/make-snippet \"note-capture_\")"))))
           :replace)

```

With snippet:

```
# -*- mode: snippet -*-
# name: note-capture
# key: note-capture_
# --
* TODO ${1:Note title}
:PROPERTIES:
:CREATED: %U
:END:
- [ ] Check what notes are followed by this
- [ ] Refile to relevant topics
- [ ] Elaborate
Following up: ${2:%a}

$0 $2

```

The capture simply exits silently if I try this capture template.

**Comment by ag91** [comment]
Hey thanks for trying it out!

I like that setup syntax, but it did not work with my Emacs setup: what package are you using? Looks cool!

So I tried to (sort of) reproduce with this setup:
```
(setq org-capture-templates
          `(("E" "Note" entry (file "/tmp/test.org") "* %(ya-org-capture/make-snippet \"note-capture_\")" :empty-lines 2 :clock-in t :clock-resume t)))
```

Also I needed to load your snippet saved in my `snippets/org-mode/note-capture_` and it seemed to work as expected (I am happy that even the time settings weren't breaking the template -- I totally did not test that).

Note the `"* %(ya-org-capture/make-snippet \"note-capture_\")"` and I mean the asterisk: with the `entry` type I needed to provide it otherwise the template was not valid. Probably fixable by using `:type plain` [in your setup](https://orgmode.org/manual/Template-elements.html#Template-elements) but I did not try.

I can give another and more precise go at reproducing this if you do not mind sharing the package you are using.

**Comment by yantar92** [comment]
> I like that setup syntax, but it did not work with my Emacs setup: what package are you using? Looks cool!

That code require `doct` [1] and `asoc` [2] packages.

I just tried your version with `*`. The capture works now, but capture template did not expand `%U`:

> * TODO blah
> :PROPERTIES:
> :CREATED:  %U
> :END:
> :LOGBOOK:
> CLOCK: [2020-07-29 Wed 20:00]
> :END:
> - [ ] Check what notes are followed by this
> - [ ] Refile to relevant topics
> - [ ] Elaborate
> Following up: asd



[1] https://github.com/progfolio/doct
[2] https://github.com/troyp/asoc.el

**Comment by ag91** [comment]
**TL;DR**: cannot expand `%U` properly yet, you can work around it using the YASnippet syntax `%(format-time-string "[%Y-%m-%d]")`

Thanks, I am going to try these packages out!
Let me make sure I get what you are trying to achieve: you would like to use a mixture of org-capture and YASnippet syntax in your snippet, is it right?
What I mean is that `%U` is org-capture syntax while `${1:Note title}` is YASnippet syntax.
I did not get there yet: now I am inserting the YASnippet keyword and expanding when the org-capture expansion is already done.
That prevents the `%U` to get expanded properly (`yas-expand` or `yankpad-expand` do not interpret that syntax).
As a workaround you can use something like `%(format-time-string "[%Y-%m-%d]")` instead of the `%U`.
I know how to solve this once and for all, but I would prefer to really need the change to work on it. 
Does the workaround work for you for now?

**Comment by yantar92** [comment]
> As a workaround you can use something like %(format-time-string "[%Y-%m-%d]") instead of the %U.

Sure, it works (with =backquote= brackets). However, the %a part is actually much more important. It inserts the link to place the capture was invoked from. I can imagine a workaround by accessing `org-capture-curret-plist` directly, but that sounds inconvinient. I proper way to handle this is probably calling `org-capture-fill-template` directly.

**Comment by ag91** [comment]
Ah that's a fair point. I did not think about
`org-capture-fill-template`: it seems simple, I will test calling
`org-capture-fill-template` after the `yas-expand` and see if that works
fine.

Another workaround is to have a mixed template: in my configuration I
have something like

```
("s" "Some Todo" entry (file "someFile.org") "* TODO %^{SomeText} :someTag:\n%i\nSee: %a\n%(ya-org-capture/make-snippet \"bi\")" :empty-lines 2)
```

Somehow I thought to have a mixed template from the beginning, and this
has biased my implementation.

**Comment by ag91** [comment]
I am afraid `org-capture-fill-template` requires to have a filled-in
snippet as a string, but that is the main obstacle I found in my first
implementation.

I will give it some thought, hopefully the workarounds will help for
now.

**Comment by yantar92** [comment]
>  ag91 commented 6 hours ago
I am afraid `org-capture-fill-template` requires to have a filled-in
snippet as a string, but that is the main obstacle I found in my first
implementation.

Maybe you can replace the buffer contents in `yas-before-expand-snippet-hook`.

**Comment by ag91** [comment]
Good idea! I tried to use `yas-after-exit-snippet-hook` instead and it
seems to work for me. The fix only works for YASnippet though.

Would you mind to try it and see if it works as
you expect?

**Comment by yantar92** [comment]
> Would you mind to try it and see if it works as you expect?

It works now. However, the expansion happens only after going through all the yas fields, which defeats the purpose of `${1:%a}` in the snippet. If `%a` is not expanded while I am filling the snippet, there is no way for me to know if I want to change `%a` to something else. 

It might be possible to achieve earlier expansion by replacing `%X` in the snippet with `(org-capture-fill-template "%X")` or something along that line. 

The following snippet achieves what I want, except looking messy:

```
# -*- mode: snippet -*-
# name: note-capture
# key: note-capture_
# --
TODO ${1:Note title}
:PROPERTIES:
:CREATED: `(format-time-string "[%Y-%m-%d]")`
:END:
- [ ] Check what notes are followed by this
- [ ] Refile to relevant topics
- [ ] Elaborate

`(org-capture-fill-template "%a")`

Following up: ${2:`(org-capture-fill-template "%a")`}

$0
```

Also,  the following function throws error when yankpad is not installed:

```
(defun ya-org-capture/snippet-expand ()
  "Try to expand snippet at point with `yankdpad-expand' and then with `yas-expand'."
  (interactive)
  (funcall (ya-org-capture/or-else 'yankpad-expand 'yas-expand) nil))

```

**Comment by ag91** [comment]
I see the issue with expanding the org-capture syntax only after, I will
see if I can come up with something better later. Anyway I think we can
close this issue because it is possible to use org-capture syntax, right?

By the way, thanks a lot for reporting! I copied your idea of defining
the whole org-mode task as a YASnippet and I think my capturing flow has
improved now.

Also I made the expand functions customizable: so if you just use
YASnippet and not Yankpad you just have to set this in your
configuration

`(setq ya-org-capture/expand-snippet-functions (list 'yas-expand))`

I left both in as default because I very like both modes.

**Comment by yantar92** [comment]
Thanks. Closing now.
